File: CivitasRuby/lib/titulo_propiedad.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module Civitas
  class TituloPropiedad
    
    @@factorInteresesHipoteca = 1.1
    
    def initialize(nom, fr, ab, hb, pc, pe)
      @nombre = nom
      @alquilerBase = ab
      @fatorRevalorizacion = fr
      @hipotecaBase = hb
      @precioCompra = pc
      @precioEdificar = pe
      @numCasas = 0
      @numHoteles = 0
      @hipotecado = false
      @propietario = nil
    end
    
    
    def to_s
      "TituloPropiedad: \n\tNombre:  #{@nombre}  \n\t
                                              Alquiler Base:  #{@alquilerBase}\n\t
                                              Factor Revalorizacion:  #{@factorRevalorizacion}\n\t
                                              Hipoteca Base:  #{@hipotecaBase}\n\t
                                              Precio Compra:  #{@precioCompra}\n\t
                                              Precio Edificar:  #{@precioEdificar}\n\t
                                              Numero de casas:  #{@numCasas}\n\t
                                              Numero de hoteles: #{@numHoteles}\n\t
                                              Hipotecado:  #{@hipotecado}  \n\t";
    end
    
    def actualiza_propietario_por_conversion(jugador)
      @propietario = jugador
    end
    
    
    def precio_alquiler
     precio_alquiler = 0
     if @hipotecado or propietarioEncarcelado
       return precio_alquiler = 0
     else
       return precio_alquiler = @alquilerBase*(1+(@numCasas*0.5) + (@numHoteles*2.5))
     end
    end
    
    def importe_cancelar_hipoteca
      result = @hipotecaBase * @@factorInteresesHipoteca
      return result
    end
    
 
    def es_este_el_propietario(jugador)
      return jugador == @propietario
    end
    
    def tramitar_alquiler(jugador)
      if(tiene_propietario && es_este_el_propietario(jugador))
        jugador.paga_alquiler(precio_alquiler)
        @propietario.recibe(precio_alquiler)
      end
    end
    

    def propietarioEncarcelado
         if @propietario != nil
           return @propietario.is_encarcelado
         else
           return false
         end
    end
    
    def cantidad_casas_hoteles
      result = @numCasas + @numHoteles
      return result
    end
   
    
    def precio_venta
      result = @precioCompra + (@numCasas+5*@numHoteles)*@precioEdificar*@factorRevalorizacion
      return result
    end
    
    def derruir_casas(n, jugador)
      if @propietario == jugador && @numCasas >= n
        @numCasas = @numCasas - n
        return true
      else
        return false
      end
    end
    
    def vender(jugador)
         if es_este_el_propietario(jugador) && !@hipotecado
           jugador.recibe(@precioCompra)
           @propietario = nil
           @numCasas = 0
           @numHoteles = 0
           return true
         else
           return false
         end
    end
    
    def tiene_propietario
      return @propietario != nil
    end
    

    def cancelar_hipteca(jugador)
      result = false
      if @hipotecado and  es_este_el_propietario(jugador)
        jugador.paga(importe_cancelar_hipoteca)
        @hipotecado = false
        result = true
      end
      return result
    end
    
    def comprar(jugador)
      result = false
      
      if !tiene_propietario
        @propietario = jugador
        result = true
        jugador.paga(precioCompra)
      end
      
      return result
    end
    
    def construir_hotel(jugador)
      result = false
      
      if es_este_el_propietario(jugador)
        jugador.paga(precioEdificar)
        @numHoteles = @numHoteles + 1
        result = true
      end
      
      return result
    end
    
    def construir_casa(jugador)
      result = false
      
      if es_este_el_propietario(jugador)
        jugador.paga(precioEdificar)
        @numCasas = @numCasas + 1
        result = true
      end
      
      return result
    end
    
    def hipotecar(jugador)
      salida = false
      
      if(!@hipotecado and es_este_el_propietario(jugador))
        jugador.recibe(hipotecaBase)
        @hipotecado = true
        salida = true
      end
      
      return salida
    end
    
    attr_reader :precioEdificar, :numCasas, :numHoteles, :hipotecado, :nombre, :precioCompra, :propietario, :hipotecaBase
    
    
  end
end

File: CivitasJava/src/civitas/TituloPropiedad.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

/**
 *
 * @author luis
 */
public class TituloPropiedad {
    private float alquilerBase;
    private static float factorInsteresesHipoteca = 1.1f;
    private float factorRevalorizacion;
    private float hipotecaBase;
    private boolean hipotecado;
    private String nombre;
    private int numCasas;
    private int numHoteles;
    private float precioCompra;
    private float precioEdificar;
    private Jugador propietario;
    
    //nombre, precio base de alquiler, factor de revalorización, precio base de hipoteca, precio de compra y precio por edificar
    TituloPropiedad(String nom, float fr, float ab, float hb, float pc, float pe){
        nombre = nom;
        alquilerBase = ab;
        factorRevalorizacion = fr;
        hipotecaBase = hb;
        precioCompra = pc;
        precioEdificar = pe;
        numCasas = 0;
        numHoteles = 0;
        hipotecado = false;
        propietario = null;
    }
    
    @Override
    public String toString(){
        String resultado = "Nombre:" + this.nombre + "\n\t"
                           + "Alquiler Base:" + this.alquilerBase + "\n\t"
                           + "Factor Revalorizacion:" + this.factorRevalorizacion + "\n\t"
                           + "Hipoteca Base:" + this.hipotecaBase + "\n\t"
                           + "Precio Compra:" + this.precioCompra + "\n\t"
                           + "Precio Edificar:" + this.precioEdificar + "\n\t"
                           + "Numero de casas:" + this.numCasas + "\n\t"
                           + "Numero de hoteles:" + this.numHoteles + "\n\t"
                           + "Hipotecado:" + this.hipotecado + "\n\t";
        
        return resultado;
    }
    
    void actualizaPropietarioPorConversion(Jugador jugador){
        this.propietario = jugador;
    }
    
    private float getPrecioAlquiler(){
        float precioAlquiler = 0;
        if(this.hipotecado || propietarioEncarcelado())
            return precioAlquiler = 0;
        else
            return precioAlquiler = this.alquilerBase*
                                    (1+(this.numCasas*0.5f)+
                                    (this.numHoteles*2.5f));
        
    }
    
    float getImporteCancelarHipoteca(){
        float result = this.hipotecaBase*this.factorInsteresesHipoteca;
        return result;
    }
    
    private boolean esEsteElPropietario(Jugador jugador){
        return jugador == propietario;
    }
    
    void tramitarAlquiler(Jugador jugador){
        if(this.tienePropietario() && esEsteElPropietario(jugador)){
            jugador.pagaAlquiler(this.getPrecioAlquiler());
            propietario.recibe(getPrecioAlquiler());
        }
    }
    
    private boolean propietarioEncarcelado(){
        if(propietario != null)
            return propietario.isEncarcelado();
        else
            return false;
    }
    
    int cantidadCasasHoteles(){
        int result = this.numCasas + this.numHoteles;
        return result;
    }
    
    private float getPrecioVenta(){
        float result = this.precioCompra + 
                       (numCasas+5*numHoteles)*precioEdificar
                       *factorRevalorizacion;
        return result;
    }
    
    boolean derruirCasas(int n, Jugador jugador){
        if(propietario == jugador && numCasas >= n){
            numCasas = numCasas - n;
            return true;
        }else
            return false;
    }
    
    boolean vender(Jugador jugador){
        if(esEsteElPropietario(jugador) && !this.hipotecado){
            jugador.recibe(this.precioCompra);
            propietario = null;
            numCasas = 0;
            numHoteles = 0;
            return true;
        }else
            return false;
    }
    
    float getPrecioEdificar(){
        return precioEdificar;
    }
    
    public int getNumCasas(){
        return numCasas;
    }
    
    public int getNumHoteles(){
        return numHoteles;
    }
    
    public boolean getHipotecado(){
        return hipotecado;
    }
    
    private float getImporteHipoteca(){
        return hipotecaBase;
    }
    
    public String getNombre(){
        return this.nombre;
    }
    
    float getPrecioCompra(){
        return this.precioCompra;
    }
    
    
    Jugador getPropietario(){
        return propietario;
    }
    
    boolean tienePropietario(){
        return propietario != null;
    }
    
    boolean cancelarHipoteca(Jugador jugador){
        boolean result = false;
        if(this.hipotecado && this.esEsteElPropietario(jugador)){
            jugador.paga(this.getImporteCancelarHipoteca());
            hipotecado = false;
            result = true;
        }
        return result;
    }
    
    boolean comprar(Jugador jugador){
        boolean result = false;
        
        if(!this.tienePropietario()){
            propietario = jugador;
            result = true;
            jugador.paga(precioCompra);
        }
        
        return result;
    }
    
    boolean construirHotel(Jugador jugador){
        boolean result = false;
        
        if(this.esEsteElPropietario(jugador)){
            jugador.paga(precioEdificar);
            numHoteles = numHoteles +1;
            result = true;
        }
        
        return result;
    }
    
    boolean construirCasa(Jugador jugador){
        boolean result = false;
        
        if(this.esEsteElPropietario(jugador)){
            jugador.paga(precioEdificar);
            numCasas = numCasas +1;
            result = true;
        }
        
        return result;
    }
    
    boolean hipotecar(Jugador jugador){
        boolean salida = false;
        
        if(!hipotecado && this.esEsteElPropietario(jugador)){
            jugador.recibe(getImporteHipoteca());
            hipotecado = true;
            salida = true;
        }
        
        return salida;
    }
    
}
